doc/compiler/osr/OSR.md

```diff
@@ -0,0 +1,276 @@
+
+Runtime assumptions are added to patch the inserted OSR guards in the event of a method redefinition,
+triggering the transition of execution to the JVM via the OSR mechanism. The OSR guards are permanently patched during an OSR yield to the JVM at which invalidation happens (e.g. a redefinition occurs). An important distinction is that all OSR guards will be patched in the event of any redefinition,
+rather than just the required set to protect a removed HCR guard. This is due to the high overhead of tracking and describing the minimal set of OSR guards that would need to be patched on invalidation of each individual assumption.
+
+## OSR and virtual guard merging
+
+As an optimization, we can mark some existing virtual guard as an OSR guard, therefore merging them together. This is achieved by making virtual guards kill fear. In this case, the taken side of a merged OSR guard will be a cold call (vs. OSR induce block). The disadvantage of guard merging is that a later type analysis that would be sufficient to remove a virtual guard won't be able to also remove the corresponding OSR guard, which likely prevents the elimination of a control flow diamond.
+
+
+## OSR induce blocks
+
+The main purpose of the OSR induce blocks is to voluntarily induce OSR.
+In addition, since OSR catch blocks apply for any possible transition points in a method, the OSR induce block allows bookkeeping to be specified at an individual OSR point.
+It contains **irrelevant stores**, which store default values to symrefs.
+These ensure that variables that are dead at a specific OSR point don't spuriously appear to be live based on uses in the OSR code block.
+Moreover, the block potentially contains rematerializaion of live symrefs and monitor enters, to further reduce the constraints on other optimizations.
+The OSR induce blocks end with an OSR induce (a call to jitInduceOSRAtCurrentPC helper) which transitions to the VM, and an exception edge to the OSR catch block.
+Once the induce helper is called, we transition to the VM.
+
+
+## Cannot attempt OSR
+
+If the JIT executes an operation via intermediate states that differ from the VM, then we cannot transition during that execution since
+the semantics of what the JIT has done cannot be expressed in terms of the semantics the VM would have executed.
+This is evident in the JIT's thunk archetype based implementation of MethodHandles.
+The calls have their BCI flagged as doNotProfile. If those calls are later inlined (which is likely), the doNotProfile bit causes the corresponding inline frames to satisfy cannotAttemptOSRDuring().
```

Comment:
I think BCI should be written out as "bytecode indices (BCIs)" here.  It might be good to add a link to a definition, if we have one someplace.

Reply:
Interestingly, could not find a definition in the openj9 docs.

Reply:
BCI here means the `TR_ByteCodeInfo` rather than bytecode index. It's really too bad that they abbreviate the same way to create this ambiguity

Reply:
Right, but I think the intent is clear  - we associate some info with BCI.